src/CloudNative.CloudEvents/Core/BinaryDataUtilities.cs

```diff
@@ -15,31 +18,96 @@ public static class BinaryDataUtilities
     {
         /// <summary>
         /// Asynchronously consumes the remaining content of the given stream, returning
-        /// it as a byte array.
+        /// it as a read-only memory segment.
         /// </summary>
         /// <param name="stream">The stream to read from. Must not be null.</param>
-        /// <returns>The content of the stream (from its original position), as a byte array.</returns>
-        public async static Task<byte[]> ToByteArrayAsync(Stream stream)
+        /// <returns>The content of the stream (from its original position), as a read-only memory segment.</returns>
+        public async static Task<ReadOnlyMemory<byte>> ToReadOnlyMemoryAsync(Stream stream)
         {
             Validation.CheckNotNull(stream, nameof(stream));
-            // TODO: Optimize if it's already a MemoryStream?
+            // TODO: Optimize if it's already a MemoryStream? Will only work in some cases,
+            // and is most likely to occur in tests, where the efficiency doesn't matter as much.
             var memory = new MemoryStream();
             await stream.CopyToAsync(memory).ConfigureAwait(false);
             return memory.ToArray();
```

Comment:
Can use MemoryStream.GetBuffer here - e.g. - https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/System.Memory.Data/src/BinaryData.cs#L179

Reply:
Except `GetBuffer` can throw `UnauthorizedException` in various cases. That's what I meant about only working in some cases :)

Reply:
Hmm good point, it looks like it will throw if constructed with the parameterless ctor, but not the one accepting a length. 
So if the stream is seekable and we construct the stream using the stream length, we should be able to use GetBuffer, which makes me think there is a bug in BinaryData.

Reply:
> That's what I meant about only working in some cases :)

I thought you were referring here to whether or not we should type sniff the stream and avoid copying. I agree that this will not always work. My suggestion is around avoiding the ToArray call, which it looks like should be possible if the original stream is seekable.

Reply:
We can use TryGetBuffer, yes. Let's revisit this later if we need to :)

Reply:
Actually I was mistaken - GetBuffer should always work here. We own the lifetime of the MemoryStream and we are using the constructorless parameter - https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream.-ctor?view=net-5.0#System_IO_MemoryStream__ctor

Reply:
Ah - I'd misinterpreted your comment. I was hoping to use TryGetBuffer on the *incoming* stream to avoid having to copy to a new stream.
Yes, I agree we can use GetBuffer here. Will do that on Monday too, in a separate PR.

Reply:
Ah that makes sense. Yes I don't recall why we didn't use TryGetBuffer on the original stream in the BinaryData implementation - it seems like a good enhancement.

Reply:
I think it may have been related to ownership of the stream - we didn't want to have to worry about the semantics of the original stream being disposed.